### DISCUSSION TITLE
Skippable return

### DISCUSSION THREAD

**Discussion by olugt** [dotnet/csharplang#4173] [Language Ideas] [answered]
UPDATE 2020-11-25 GMT+1:
For the sake of clarity, the main objective is how to return from a method without having to type lots of `if` code. It's not about handling error. Also, it is not about out parameter because an async method cannot use an out parameter. Hence, the "skippable return" idea. If tuple is not ideal, contributors should kindly suggest any viable alternatives.

MAIN DISCUSSION:
Please, in C#, can we have something like the following...

    skippable return ProcessExists(username);

    skippable return ProcessUnlockStatus(username, someOtherParameter);

    return null;

...instead of...

    var (doesUserExist, message_UserExistence) = ProcessExists(username);
    if (!doesUserExist)
        return message_UserExistence;

    var (isUserUnlocked, message_UserUnlockStatus) = ProcessUnlockStatus(username, someOtherParameter);
    if (!isUserUnlocked)
        return message_UserUnlockStatus;

    return null;

...where...

    skippable return

...expects a tuple in the form of...

    (bool someBoolean, object someObjectToBeReturned)

...and in my example, it is such that if isOk is false, the method should return. But that's just my use case. Another person may prefer to return when isOk is true. So, it depends on which way the eventual spec would choose.

Please, I really think it's great.

I had tweeted about the original text (already modified here) at...
https://mobile.twitter.com/OluGt_/status/1331292170230312960

**Comment by olugt** [comment]
The code is not intuitive but I had to come up with something to demonstrate what I am trying to say and keep the narration as short as possible.

**Reply by olugt** [reply]
I mean that the valid C# code is the way it is in the idea just for the sake of describing the idea. I'm aware the method names or intents may not sit well with a lot of people. But I had to type something as short as possible.

**Comment by HaloFour** [comment, marked as answer]
See: https://github.com/dotnet/csharplang/issues/3242

**Reply by olugt** [reply]
Not same. I traced the other discussions too, but not same.

**Reply by HaloFour** [reply]
It's similar.

Either way, my opinion is that the "expanded" form is significantly easier to read.

You also have the option of following the idiomatic `Try*` signature for these kinds of methods:

```cs
if (ProcessExists(username, out var message_Ex)) return message_Ex;
if (ProcessLock(username, someOtherParameter, out var message_Lock)) return message_Lock;
return null;
```

**Reply by olugt** [reply]
The out parameter is limiting. Not all methods can use out parameter.

**Reply by olugt** [reply]
Also, your suggestion leads to more typing. Look at my idea again...

    skippable return ProcessExists(username);

**Reply by olugt** [reply]
Btw, your suggestion is quite analogous to returning an actual class vs tuple from a method. Tuple makes it quick and concise when there's really no need to create a class. Similar to record type vs actual full-fledged class.

**Reply by HaloFour** [reply]
The amount of typing isn't that important.  The question is clarity.  In my opinion it's not at all clear what your code does.  IMO, the original code you've written is substantially easier to read.  I don't like the idea of a single statement that applies either a conditional or control flow based on some implicit interpretation of the result of an expression.  The use of tuples in that manner is not terribly common in C# and the `bool/out` signature is considered more idiomatic and is already pretty concise.

**Reply by olugt** [reply]
Async methods can't use out parameter.

**Reply by olugt** [reply]
Check the update I made to the idea, because some people thought it's about error handling.

**Reply by olugt** [reply]
The boolean (the first parameter in the tuple) does only one thing, which is to determine whether or not the method should return immediately right there. The second parameter can be any kind of object (not just string as used in my example) because throwing an exception is not always applicable when something doesn't go in a particular way, as an object may want to be returned instead. So, again, the boolean is strictly for determining whether to return or not.

**Reply by HaloFour** [reply]
You can also use pattern matching here:

```cs
if (ProcessExists(username) is (true, var message_Ex)) return message_Ex;
if (ProcessLock(username) is (true, var message_Lock)) return message_Lock;
return;
```

And this does work with `await`.

**Reply by olugt** [reply]
I AM SHOCKED❗❗❗

That is the answer❗❗❗

That is why I like to be understood. Thank you for staying with the argument and not giving up❗

I have checked it with an async method and it works❗❗❗

**Reply by olugt** [reply]
Please, I can't seem to get the URL to this answer. If you can, kindly respond with a URL to this answer so that I can post it on the tweets thread on Twitter, quoting your answer.

**Reply by canton7** [reply]
The link is on the text "17h ago" (or whatever the time is): https://github.com/dotnet/csharplang/discussions/4173#discussioncomment-138042

**Reply by olugt** [reply]
Thank you so much❗

**Reply by olugt** [reply]
Thank you so much, again, @HaloFour ❗❗❗ I should not forget to say it works for pattern-matching types too, just like anything else❗

    if (ProcessSomething(username) is { Property1: 123, Property2: var prop2Value }) return prop2Value;

    if (ProcessSomethingElse(username) is { PropertyBle: var propBleValue, PropertyBla: 456 }) { logger.LogSomething(propBleValue); return propBleValue; };
    
    return null;

**Comment by olugt** [comment]
It should be noted that other scenarios of the use of `return` can cause modification to the idea.

For example, if no return value is applicable, making the return skippable can be by just supplying a boolean value like...

    skippable return SomeBoleanCheck(username);

    return;

**Comment by canton7** [comment]
From the snippet you posted, it looks you're doing error handling by passing bools around. If this is the case, the idiomatic way to do this in .NET would be using exceptions. Then you can write:

```cs
EnsureProcessExists(username);
LockProcess(username, someOtherParameter);
```

The functional approach would be to use a `Result<TResult, TError>` type instead of a tuple. There are some other discussions around adding some capabilities to the language to let you return if a `Result<TResult, TError>` contains an error in a syntax-efficient way (similar to e.g. Rust's `?` or Swift's `try`): you'll have more luck adding your voice to those, although I don't think they'll get very far very soon.

**Reply by olugt** [reply]
I'm not passing error.

It only describes a scenario where I can return without using lots of guard code. So, maybe the use of tuple makes it somehow. But the main objective is how to return from a method without having to type lots of if statements. Hence, "skippable return".

**Reply by olugt** [reply]
Check the update I made to the idea, because some people thought it's about error handling.

**Reply by canton7** [reply]
My point about it being a bad version of Result still stands; go find one of those proposals.

**Reply by olugt** [reply]
The boolean (the first parameter in the tuple) does only one thing, which is to determine whether or not the method should return immediately right there. The second parameter can be any kind of object (not just string as used in my example) because throwing an exception is not always applicable when something doesn't go in a particular way, as an object may want to be returned instead. So, again, the boolean is strictly for determining whether to return or not.

Btw, I tried to search for the discussion you referred to but no search result came up. Also, it is not same as my idea, as I just explained.

**Reply by canton7** [reply]
I think it is the same as your idea. A `Result<TResult, Error>` contains a bool saying whether the operation succeeded. If it did succeed, it contains an (optional) result. If it failed, it contains a resulting error. Languages which use `Result<TResult, TError>` often have syntax to return straight away with the error if a result failed. So your idea *is* equivalent to a `Result` + syntax to return with the error part of the result if the result failed I think: it just uses a union type rather than a tuple.

Discussion on `Result` happened in the various disctriminated union proposals. There's a discussion on the "return if a result failed" bit [here](https://github.com/dotnet/csharplang/discussions/3403).

**Comment by canton7** [comment]
Also:

> I tweeted about this at...
> https://mobile.twitter.com/OluGt_/status/1331292170230312960

[Mads replied](https://mobile.twitter.com/MadsTorgersen/status/1331383732599549952):

> I'm not a fan. It's too much of a special case to deserve its own syntax. Personally I think the Try...(out var ...) pattern does a good job here, and is much more general.

**Reply by olugt** [reply]
Nevertheless, that's why it's an idea. He simply expressed his view.
